python/descriptor_containers.c

```diff
@@ -0,0 +1,670 @@
+  } else {
+    PyErr_SetObject(PyExc_KeyError, key);
+    return NULL;
+  }
+}
+
+static int PyUpb_ByNameMap_AssignSubscript(PyObject *self, PyObject *key,
+                                           PyObject *value) {
+  PyErr_Format(PyExc_TypeError, PYUPB_MODULE_NAME
+               ".ByNameMap' object does not support item assignment");
+  return -1;
+}
+
+static int PyUpb_ByNameMap_Contains(PyObject *_self, PyObject *key) {
+  PyUpb_ByNameMap *self = (void*)_self;
+  const char *name = PyUpb_GetStrData(key);
+  const void *elem = name ? self->funcs->lookup(self->parent, name) : NULL;
+  return elem ? 1 : 0;
+}
+
+static PyObject *PyUpb_ByNameMap_Get(PyObject *_self, PyObject *args) {
+  PyUpb_ByNameMap *self = (void*)_self;
+  PyObject* key;
+  PyObject* default_value = Py_None;
+  if (!PyArg_UnpackTuple(args, "get", 1, 2, &key, &default_value)) {
+    return NULL;
+  }
+
+  const char *name = PyUpb_GetStrData(key);
+  const void *elem = name ? self->funcs->lookup(self->parent, name) : NULL;
+
+  if (elem) {
+    return self->funcs->base.get_elem_wrapper(elem);
+  } else {
+    Py_INCREF(default_value);
+    return default_value;
+  }
+}
+
+static PyObject *PyUpb_ByNameMap_GetIter(PyObject *_self) {
+  PyUpb_ByNameMap *self = (PyUpb_ByNameMap *)_self;
+  return PyUpb_DescriptorIterator_New(&self->funcs->base, self->parent,
+                                      self->parent_obj);
+}
+
+static PyObject *PyUpb_ByNameMap_Keys(PyObject *_self, PyObject *args) {
+  PyUpb_ByNameMap *self = (PyUpb_ByNameMap *)_self;
+  int n = self->funcs->base.get_elem_count(self->parent);
+  PyObject *ret = PyList_New(n);
+  if (!ret) return NULL;
+  for (int i = 0; i < n; i++) {
+    const void *elem = self->funcs->base.index(self->parent, i);
+    PyObject *key = PyUnicode_FromString(self->funcs->get_elem_name(elem));
+    if (!key) return NULL;
+    PyList_SetItem(ret, i, key);
+  }
+  return ret;
+}
+
+static PyObject *PyUpb_ByNameMap_Values(PyObject *_self, PyObject *args) {
+  PyUpb_ByNameMap *self = (PyUpb_ByNameMap *)_self;
+  int n = self->funcs->base.get_elem_count(self->parent);
+  PyObject *ret = PyList_New(n);
+  if (!ret) return NULL;
+  for (int i = 0; i < n; i++) {
+    const void *elem = self->funcs->base.index(self->parent, i);
+    PyObject *py_elem = self->funcs->base.get_elem_wrapper(elem);
+    if (!elem) return NULL;
+    PyList_SetItem(ret, i, py_elem);
+  }
+  return ret;
+}
+
+static PyObject *PyUpb_ByNameMap_Items(PyObject *_self, PyObject *args) {
+  PyUpb_ByNameMap *self = (PyUpb_ByNameMap *)_self;
+  int n = self->funcs->base.get_elem_count(self->parent);
+  PyObject *ret = PyList_New(n);
+  if (!ret) return NULL;
+  for (int i = 0; i < n; i++) {
+    PyObject *item = PyTuple_New(2);
+    if (!item) return NULL;
+    const void *elem = self->funcs->base.index(self->parent, i);
+    if (!elem) return NULL;
+    PyObject *py_elem = self->funcs->base.get_elem_wrapper(elem);
+    if (!py_elem) return NULL;
```

Comment:
pretty sure you are leaking `item` and `ret` on the early returns here

Reply:
Done.